File: src/domains/biosphere/perturbations.py
"""Perturbation harness (Phase-3 Step 6, P3.5) — forcing + flow interventions.

A perturbation is a **scenario-layer intervention composed onto the already-assembled
``(state, registry, resolver)``**, not a core change. Each shipped perturbation shows a
**cascade with no cascade code** while **conservation + ``rationed == 0``** hold through
it (P3.5). The harness composes *outside* ``build_season``, so the three existing
goldens and ``src/simcore/`` are untouched (``git diff src/simcore/`` stays empty); a
perturbation run gets **no golden** (it is a behavioural demonstration — Phase 4 owns
freeze-as-reference; the Step-5 "diagnostics, no golden" precedent).

Every perturbation reduces to a **pure function of the integer step ``n``** — the
legitimate forcing seam (#14: schedules take ``(n, dt)``, evaluated at ``t = n·dt``).
**Two seam-types, both already legitimate:**

1. **Forcing perturbation** (drought, lighting failure) — a pure schedule transform that
   wraps **one** forcing var's :data:`~simcore.environment.Schedule` with a windowed
   override and rebuilds the resolver with that var replaced (:func:`with_forcing` +
   :func:`window_override`). **No new stock / flow / sink; the ledger structure is
   untouched** (only a forcing *value* changes over a window), so conservation is
   structurally unaffected. :func:`with_drought` cuts irrigation to 0; the open field is
   the only scenario with irrigation to cut (the sealed chamber dropped it in Step 3 for
   genuine water closure). :func:`with_lighting_failure` cuts PAR to 0.

2. **Flow perturbation** (atmospheric leak) — :func:`with_atmospheric_leak` augments
   ``(state, registry)`` with a :class:`LeakFlow` + a boundary leak-sink stock, the
   leak's **timing gated by a windowed forcing var** read via ``env.get`` (so the
   calendar lives in a *schedule*, not the rate law — honouring the Step-4 "no calendar
   in a flow body" discipline; the flow itself is a pure first-order rate law). The leak
   var (:data:`LEAK_VAR`) is **local to this module, never added to the ``stocks.py``
   catalog** — baseline assembly must never see it, or the existing goldens are touched.

**Rejected — a ``Perturbation`` protocol / dataclass with
``.apply(state, registry, resolver)``.** Speculative generality for 2–3 perturbations
(the same critique that killed the Step-5 observe-hook): the three small builder
functions + the two shared helpers here are leaner and add no surface for a consumer
that does not exist. A protocol later is additive if a real multi-perturbation
composition need appears.

**Recoverable-regime constraint (a Step-6 probe finding).** Graceful cascades require a
window **within one year** and a magnitude bounded so the plant refills its grain
(``storage_c``) before the next annual reset. A *severe or permanent* perturbation
suppresses photosynthesis enough that the grain never refills, and at the year boundary
:func:`~domains.biosphere.season.annual_reset` **raises ``ValueError`` (seed bank too
small to re-sow)** — a driver crash, not a cascade, and arguably correct (a sealed
chamber whose plant cannot rebuild its seed bank genuinely cannot re-sow). The shipped
perturbations stay transient; the regime boundary is pinned by a characterization test,
not grown into scope.

Pure stdlib + ``simcore`` + ``domains`` (no builder imports another, P3.3).
"""

from dataclasses import dataclass, replace

from domains.biosphere.stocks import IRRIGATION_VAR, PAR_VAR
from simcore.boundary import BOUNDARY_DOMAIN
from simcore.environment import (
    Environment,
    Schedule,
    SourceResolver,
    constant,
)
from simcore.flow import FlowResult, Leg
from simcore.ids import FlowId, StockId
from simcore.quantities import StockKind
from simcore.registry import Registry
from simcore.state import State, Stock

# The leak's windowed-activation forcing var (1.0 inside the window, 0.0 outside).
# **Local to this module** — never added to the ``stocks.py`` catalog, so the baseline
# season/resolver never carry it and the three existing goldens stay byte-identical.
LEAK_VAR = "atmospheric_leak"
# The boundary reservoir the leak vents into (boundary-domain, so the chamber's interior
# closure breaks but total mass — interior + this sink — is conserved). Also local.
LEAK_SINK: StockId = StockId("boundary.leak_sink")


# --- the two shared helpers (forcing-schedule perturbations) -----------------


def window_override(base: Schedule, *, start: int, end: int, value: float) -> Schedule:
    """A ``Schedule`` forcing ``value`` on ``[start, end)``, else the wrapped ``base``.

    The override is a pure function of the integer step ``n`` (#14) — the legitimate
    forcing seam: inside ``[start, end)`` it forces ``value`` (e.g. irrigation/PAR → 0,
    or a leak's activation → 1), outside it defers to the wrapped ``base`` schedule
    (the unperturbed forcing). ``dt`` is threaded to ``base`` unchanged.
    """

    def schedule(n: int, dt: float) -> float:
        if start <= n < end:
            return value
        return base(n, dt)

    return schedule


def with_forcing(
    resolver: SourceResolver, var: str, schedule: Schedule
) -> SourceResolver:
    """Rebuild ``resolver`` with one forcing ``var`` replaced/added; shared map kept.

    The structural complement to :func:`window_override`: the override changes a *value*
    over a window, this swaps the *wiring* for one var and rebuilds the immutable
    resolver. The ``shared`` map (the #16 live-stock seam) is copied verbatim, so the
    forcing/shared disjointness ``SourceResolver`` enforces is preserved (a new forcing
    var like :data:`LEAK_VAR` must not collide with a shared var — it does not).
    """
    return SourceResolver(
        forcings={**dict(resolver.forcings), var: schedule},
        shared=dict(resolver.shared),
    )


def with_drought(resolver: SourceResolver, *, start: int, end: int) -> SourceResolver:
    """Cut irrigation to 0 over ``[start, end)`` (the drought perturbation; open field).

    Wraps the irrigation forcing with a windowed ``value=0`` override: over the window
    the soil-water pool stops refilling and transpiration draws it across the stress
    band, so ``f_water < 1`` and assimilation falls — a cascade, no cascade code. The
    open field is the only scenario with an irrigation forcing to cut (sealed dropped it
    in Step 3); on a sealed resolver this raises ``KeyError`` at the lookup, by design.
    """
    base = resolver.forcings[IRRIGATION_VAR]
    return with_forcing(
        resolver, IRRIGATION_VAR, window_override(base, start=start, end=end, value=0.0)
    )


def with_lighting_failure(
    resolver: SourceResolver, *, start: int, end: int
) -> SourceResolver:
    """Cut PAR to 0 over ``[start, end)`` (the lighting-failure perturbation).

    Wraps the PAR forcing with a windowed ``value=0`` override: over the window FvCB's
    light term collapses (``J → 0``), assimilation → 0, growth stalls, O₂ production
    drops and — in the sealed chamber — the carbon pool stops drawing down, so chamber
    CO₂ *rises*. A cascade with no cascade code; conservation is structurally unaffected
    (only a forcing value changes).
    """
    base = resolver.forcings[PAR_VAR]
    return with_forcing(
        resolver, PAR_VAR, window_override(base, start=start, end=end, value=0.0)
    )


# --- the flow perturbation (atmospheric leak) --------------------------------


@dataclass(frozen=True)
class LeakFlow:
    """A windowed boundary leak ``pool -> sink``, first-order in the pool (P3.5).

    The pure rate law is first-order donor control ``k_leak · pool · active(n) · dt``
    where ``active(n) = env.get(leak_var) ∈ {0, 1}`` carries the calendar (the window
    lives in the *schedule*, not the flow body — the Step-4 discipline). Structural
    positivity: the draw → 0 as the pool → 0, and ``k_leak·dt < 1`` keeps the Euler
    backstop unfired (``rationed == 0``) with no ``max(0, …)`` clamp — the decomposition
    / condensation precedent. ``sink`` mirrors ``pool``'s element composition (built by
    :func:`with_atmospheric_leak`), so the leg is per-quantity balanced and the gate
    folds it exactly (a CO₂ pool vents both CARBON and OXYGEN, 2:1). ``flux = daily·dt``
    (dt-linear), so the RK4 increment-form contract holds.
    """

    id: FlowId
    priority: int
    pool: StockId
    sink: StockId
    leak_var: str
    k_leak: float

    def evaluate(self, snapshot: State, env: Environment, dt: float) -> FlowResult:
        active = env.get(self.leak_var)
        amount = self.k_leak * snapshot.stocks[self.pool].amount * active * dt
        return FlowResult(legs=(Leg(self.pool, -amount), Leg(self.sink, amount)))


def with_atmospheric_leak(
    state: State,
    registry: Registry,
    resolver: SourceResolver,
    *,
    pool: StockId,
    k_leak: float,
    start: int,
    end: int,
) -> tuple[State, Registry, SourceResolver]:
    """Augment the assembled inputs with a windowed leak ``pool -> leak_sink``.

    Builds three things and returns the augmented triple (the baseline inputs are
    untouched — this composes onto them):

    * a :data:`LEAK_SINK` BOUNDARY stock whose composition **mirrors ``pool``'s**
      (so a multi-quantity CO₂ pool vents CARBON+OXYGEN in balance; built directly, not
      via ``boundary.sink`` which is single-quantity — the multi-quantity BOUNDARY is
      legal, only POPULATION is single-quantity-constrained);
    * a :class:`LeakFlow` ``pool -> leak_sink`` appended to the registry (Registry
      re-sorts by id, so append order is inert; the existing aux processes are carried);
    * the :data:`LEAK_VAR` activation forcing wired into the resolver as a windowed
      override (``1`` on ``[start, end)``, ``0`` else) — so the leak's calendar is a
      schedule, read by the flow via ``env.get``.

    The chamber's interior closure breaks over the window (mass leaves to the boundary
    leak-sink) but **total** mass (interior + leak-sink) stays conserved — the leg is
    balanced, the gate folds it. ``k_leak·dt < 1`` keeps ``rationed == 0``.
    """
    pool_stock = state.stocks[pool]
    leak_sink = Stock(
        id=LEAK_SINK,
        domain=BOUNDARY_DOMAIN,
        quantity=pool_stock.quantity,
        unit=pool_stock.unit,
        amount=0.0,
        kind=StockKind.BOUNDARY,
        composition=dict(pool_stock.composition),
    )
    new_stocks = {**dict(state.stocks), LEAK_SINK: leak_sink}
    new_state = replace(state, stocks=new_stocks)
    leak = LeakFlow(
        FlowId("biosphere.atmospheric_leak"),
        0,
        pool=pool,
        sink=LEAK_SINK,
        leak_var=LEAK_VAR,
        k_leak=k_leak,
    )
    new_registry = Registry(
        [*registry.flows, leak], new_stocks, aux_processes=registry.aux_processes
    )
    new_resolver = with_forcing(
        resolver,
        LEAK_VAR,
        window_override(constant(0.0), start=start, end=end, value=1.0),
    )
    return new_state, new_registry, new_resolver

File: rust/crates/domains/src/biosphere/perturbations.rs
//! Phase-8 (P8.5) — the **generic** perturbation primitives, the Rust port of the
//! shared half of `domains.biosphere.perturbations`.
//!
//! A perturbation is a **scenario-layer intervention composed onto the already-assembled
//! `(state, registry, resolver)`**, never a core/domain change (the Phase-3 discipline,
//! carried cross-domain by the sibling `station::perturbations`). This module holds the
//! two seam-types that are domain-agnostic — a **forcing-schedule transform**
//! ([`window_override`] + [`with_forcing`]) and an **added boundary [`LeakFlow`]** — which
//! the station composers reuse verbatim. The station-specific pieces (`ScaledFlow`, the
//! five `with_*` composers) live in `station::perturbations`.
//!
//! **The move-pipeline shape (Rust, not Python).** Python's `with_forcing` shallow-copies
//! the resolver's dict of callables; Rust cannot — a [`Schedule`] is a non-`Clone`
//! `Box<dyn Fn>`. So every forcing perturbation here *consumes* the [`SourceResolver`]
//! (via [`SourceResolver::into_parts`]), swaps one var's schedule, and rebuilds. This is
//! safe because a perturbation is applied **before** the session takes ownership — the
//! same "compose onto assembled inputs, then run" order as Python.
//!
//! **Zero parity concern.** These are diagnostic interventions (the Phase-3 "diagnostics,
//! no golden" precedent) — Phase 7 deliberately excluded them, and no golden pins a
//! perturbed run. Determinism (a re-run is bit-identical) is the no-golden insurance.

use simcore::environment::{Environment, Schedule, SourceResolver};
use simcore::error::SimError;
use simcore::flow::{Flow, FlowResult, Leg};
use simcore::ids::{FlowId, StockId};
use simcore::state::State;

/// The leak's windowed-activation forcing var (1.0 inside the window, 0.0 outside).
/// **Local to the perturbation harness** — never added to any domain's stock/forcing
/// catalog, so a baseline assembly never carries it and the frozen goldens are untouched
/// (the Python `LEAK_VAR` discipline).
pub const LEAK_VAR: &str = "atmospheric_leak";

/// The boundary reservoir a [`LeakFlow`] vents into (boundary-domain, so the interior's
/// closure breaks over the window but total mass — interior + this sink — is conserved).
/// Also local to the harness. The composer builds the actual stock with a composition
/// mirroring the leaked pool's, so a `{Carbon:1, Oxygen:2}` CO₂ pool vents both quantities.
pub const LEAK_SINK: &str = "boundary.leak_sink";

// --- the forcing-schedule perturbation (no new flow / no structural change) -----------

/// A [`Schedule`] forcing `value` on `[start, end)`, else the wrapped `base` — the Rust
/// port of Python `window_override`.
///
/// A pure function of the integer step `n` (#14, the legitimate forcing seam): inside the
/// window it forces `value` (e.g. PAR → 0, or a leak's activation → 1); outside it defers
/// to the wrapped `base` schedule. `base` is **owned** (moved into the closure) — Rust
/// cannot borrow-and-copy a `Box<dyn Fn>`, so the wrapper takes ownership. `dt` threads to
/// `base` unchanged.
pub fn window_override(base: Schedule, start: u64, end: u64, value: f64) -> Schedule {
    Box::new(move |n, dt| if start <= n && n < end { value } else { base(n, dt) })
}

/// Rebuild `resolver` with one forcing `var` replaced/added; the shared map preserved —
/// the Rust port of Python `with_forcing`.
///
/// Consumes the resolver ([`SourceResolver::into_parts`]), inserts `schedule` under `var`,
/// and rebuilds via [`SourceResolver::new`] (which re-checks the forcing⊕shared
/// disjointness, #16, so a new forcing var must not collide with a shared var — the return
/// is `Result` where Python's is infallible, the disjointness re-check being a Rust
/// feature not a Python one). The `shared` map is carried verbatim.
pub fn with_forcing(
    resolver: SourceResolver,
    var: &str,
    schedule: Schedule,
) -> Result<SourceResolver, SimError> {
    let (mut forcings, shared) = resolver.into_parts();
    forcings.insert(var.to_string(), schedule);
    SourceResolver::new(forcings, shared)
}

// --- the added leak flow (a boundary sink) --------------------------------------------

/// A windowed boundary leak `pool -> sink`, first-order in the pool — the Rust port of
/// Python `LeakFlow`.
///
/// The rate law is first-order donor control `k_leak · pool · active(n) · dt` where
/// `active(n) = env.get(leak_var) ∈ {0, 1}` carries the calendar (the window lives in the
/// *schedule*, not the flow body — the Step-4 discipline). Structural positivity: the draw
/// → 0 as the pool → 0, and `k_leak·dt < 1` keeps the Euler backstop unfired
/// (`rationed == 0`) with no clamp. `sink` mirrors `pool`'s element composition (built by
/// the composer), so the leg is per-quantity balanced. `flux = k·pool·active·dt` is
/// dt-linear (the RK4 increment-form contract, though the biosphere is Euler-locked).
pub struct LeakFlow {
    id: FlowId,
    priority: i64,
    pool: StockId,
    sink: StockId,
    leak_var: String,
    k_leak: f64,
}

impl LeakFlow {
    /// Construct a leak `pool -> sink`, gated by `leak_var`, rate `k_leak` (per second).
    pub fn new(
        id: FlowId,
        priority: i64,
        pool: StockId,
        sink: StockId,
        leak_var: String,
        k_leak: f64,
    ) -> Self {
        LeakFlow {
            id,
            priority,
            pool,
            sink,
            leak_var,
            k_leak,
        }
    }
}

impl Flow for LeakFlow {
    fn id(&self) -> &str {
        &self.id
    }

    fn priority(&self) -> i64 {
        self.priority
    }

    fn evaluate(
        &self,
        snapshot: &State,
        env: &dyn Environment,
        dt: f64,
    ) -> Result<FlowResult, SimError> {
        let active = env.get(&self.leak_var)?;
        // Character-for-character with Python: k_leak * pool.amount * active * dt.
        let amount = self.k_leak * snapshot.stocks[&self.pool].amount * active * dt;
        FlowResult::new(vec![
            Leg::new(self.pool.clone(), -amount)?,
            Leg::new(self.sink.clone(), amount)?,
        ])
    }
}

#[cfg(test)]
mod tests {
    use std::collections::{BTreeMap, HashMap};

    use simcore::environment::constant;
    use simcore::quantities::{Quantity, StockKind};
    use simcore::state::Stock;

    use super::*;

    fn pool(id: &str, amount: f64) -> Stock {
        Stock::new(
            id.to_string(),
            "d".to_string(),
            Quantity::Carbon,
            "mol".to_string(),
            amount,
            StockKind::Pool,
            0.0,
            false,
            BTreeMap::new(),
        )
        .unwrap()
    }

    #[test]
    fn window_override_forces_inside_defers_outside() {
        let base = constant(3.0).unwrap();
        let sched = window_override(base, 2, 5, 0.0);
        assert_eq!(sched(1, 1.0), 3.0); // before → base
        assert_eq!(sched(2, 1.0), 0.0); // start (inclusive) → value
        assert_eq!(sched(4, 1.0), 0.0); // inside → value
        assert_eq!(sched(5, 1.0), 3.0); // end (exclusive) → base
    }

    #[test]
    fn with_forcing_swaps_one_var_preserves_the_rest() {
        let mut forcings: HashMap<String, Schedule> = HashMap::new();
        forcings.insert("a".to_string(), constant(1.0).unwrap());
        forcings.insert("b".to_string(), constant(2.0).unwrap());
        let resolver = SourceResolver::new(forcings, HashMap::new()).unwrap();
        let resolver = with_forcing(resolver, "a", constant(9.0).unwrap()).unwrap();
        let state = State::new(0, BTreeMap::new(), 0, BTreeMap::new()).unwrap();
        let bound = resolver.bind(&state, 1.0);
        assert_eq!(bound.get("a").unwrap(), 9.0); // swapped
        assert_eq!(bound.get("b").unwrap(), 2.0); // preserved
    }

    #[test]
    fn leak_flow_is_first_order_gated_and_balanced() {
        let stocks = BTreeMap::from([
            ("p".to_string(), pool("p", 50.0)),
            (LEAK_SINK.to_string(), pool(LEAK_SINK, 0.0)),
        ]);
        let state = State::new(3, stocks, 0, BTreeMap::new()).unwrap();
        let leak = LeakFlow::new(
            "test.leak".to_string(),
            0,
            "p".to_string(),
            LEAK_SINK.to_string(),
            LEAK_VAR.to_string(),
            1.0e-3,
        );
        // active = 1 → the leak flows.
        let mut on: HashMap<String, Schedule> = HashMap::new();
        on.insert(LEAK_VAR.to_string(), constant(1.0).unwrap());
        let res_on = SourceResolver::new(on, HashMap::new()).unwrap();
        let out = leak
            .evaluate(&state, &res_on.bind(&state, 60.0), 60.0)
            .unwrap();
        // k·pool·active·dt = 1e-3·50·1·60 = 3.0 out of the pool, into the sink (balanced).
        assert_eq!(out.legs[0], Leg::new("p".to_string(), -3.0).unwrap());
        assert_eq!(out.legs[1], Leg::new(LEAK_SINK.to_string(), 3.0).unwrap());
        // active = 0 (outside the window) → the leak is inert.
        let mut off: HashMap<String, Schedule> = HashMap::new();
        off.insert(LEAK_VAR.to_string(), constant(0.0).unwrap());
        let res_off = SourceResolver::new(off, HashMap::new()).unwrap();
        let inert = leak
            .evaluate(&state, &res_off.bind(&state, 60.0), 60.0)
            .unwrap();
        assert_eq!(inert.legs[0].amount, 0.0);
        assert_eq!(inert.legs[1].amount, 0.0);
    }
}
